### PR TITLE
[MIG] stock_reserve_sale 15.0

### PR DESCRIPTION
**Pull request by flachica**
Dependencies:

https://github.com/OCA/stock-logistics-warehouse/pull/1485

### CODE REVIEW

**Comment by victoralmau** [comment]
Please, cherry-pick https://github.com/OCA/stock-logistics-warehouse/pull/1489

**Comment by flachica** [comment]
@victoralmau Forgive my ignorance, how do I apply the cherry-pick?

**Comment by flachica** [comment]
@victoralmau I think that I included your #1489 correctly

**Comment by victoralmau** [comment]
> @victoralmau I think that I included your #1489 correctly

Ok, it's well done. 
When the `stock_reserve` addon is merged, we can continue with this migration.

**Comment by rousseldenis** [comment]
@flachica  FYI, you can, if you want to tests succeed, add a line in test-requirements.txt file with the addon this PR depends on.

Like : 

`odoo-addon-stock-reserve @ git+https://github.com/OCA/stock-logistics-warehouse.git@refs/pull/1484/head#subdirectory=setup/stock_reserve`

That commit should be then removed after the other PR merging.

**Comment by flachica** [comment]
> @flachica FYI, you can, if you want to tests succeed, add a line in test-requirements.txt file with the addon this PR depends on.
> 
> Like :
> 
> `odoo-addon-stock-reserve @ git+https://github.com/OCA/stock-logistics-warehouse.git@refs/pull/1484/head#subdirectory=setup/stock_reserve`
> 
> That commit should be then removed after the other PR merging.

Bad news. [Tests](https://github.com/OCA/stock-logistics-warehouse/actions/runs/3226311049/jobs/5279674926) was failing

File: test-requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+odoo-addon-stock-reserve @ git+https://github.com/OCA/stock-logistics-warehouse.git@refs/pull/1484/head#subdirectory=setup/stock_reserve
```
Comment (rousseldenis) [inline test-requirements.txt]:
```suggestion
odoo-addon-stock-reserve @ git+https://github.com/OCA/stock-logistics-warehouse.git@refs/pull/1485/head#subdirectory=setup/stock_reserve
```

Link to good PR should work :smiley:

Comment (flachica) [reply test-requirements.txt]: Still failling. I tried your suggestion and other change but I can't solve [this](https://github.com/OCA/stock-logistics-warehouse/actions/runs/3271911843/jobs/5382272803#step:3:17).

Thanks

Comment (rousseldenis) [reply test-requirements.txt]: Please use suggested syntax. Yours is not correct

Comment (flachica) [reply test-requirements.txt]: Thanks for your response but, as I commented before, I tryied your suggestion but [still](https://github.com/OCA/stock-logistics-warehouse/actions/runs/3280454514/jobs/5401247922#step:3:15) failing

I don't find any example. Sorry. Where I can find documentation to read about this?

Comment (rousseldenis) [reply test-requirements.txt]: @flachica Not the same error.

Here this is about your module tests. Not not found module.

Comment (rousseldenis) [reply test-requirements.txt]: The error about unreleased dependencies is normal. It should be solved before merging this (as the base module is merged).

Comment (rousseldenis) [reply test-requirements.txt]: The difference here is you have a runboat running that allows to do functional tests

Comment (flachica) [reply test-requirements.txt]: Wow, impressive. I love the functionality. Thanks so much for the help

File: stock_reserve_sale/model/sale.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+                    is_stock_reservable = True
+            if sale.state not in ("draft", "sent"):
+                is_stock_reservable = False
+            sale.is_stock_reservable = is_stock_reservable
+            sale.has_stock_reservation = has_stock_reservation
+
+    has_stock_reservation = fields.Boolean(
+        compute="_compute_stock_reservation",
+        readonly=True,
+        multi="stock_reservation",
```
Comment (CasVissers-360ERP) [inline stock_reserve_sale/model/sale.py:33]: Field sale.order.has_stock_reservation: unknown parameter 'multi', if this is an actual parameter you may want to override the method _valid_field_parameter on the relevant model in order to allow it

File: stock_reserve_sale/model/sale.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        store=True,
+        string="Has Stock Reservations",
+    )
+    is_stock_reservable = fields.Boolean(
+        compute="_compute_stock_reservation",
+        readonly=True,
+        multi="stock_reservation",
```
Comment (CasVissers-360ERP) [inline stock_reserve_sale/model/sale.py:40]: Field sale.order.is_stock_reservable: unknown parameter 'multi', if this is an actual parameter you may want to override the method _valid_field_parameter on the relevant model in order to allow it

File: stock_reserve_sale/wizard/sale_stock_reserve.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
+# Copyright 2013 Camptocamp SA - Guewen Baconnier
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import _, api, exceptions, fields, models
+
+
+class SaleStockReserve(models.TransientModel):
+    _name = "sale.stock.reserve"
```
Comment (CasVissers-360ERP) [inline stock_reserve_sale/wizard/sale_stock_reserve.py:7]: You are not allowed to access 'Sale Stock Reserve' (sale.stock.reserve) records. 

Missing access rights

File: stock_reserve_sale/view/sale.xml [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            <button name="action_quotation_send" position="before">
+                <field name="is_stock_reservable" invisible="1" />
+                <button
+                    name="%(action_sale_stock_reserve)d"
+                    type="action"
+                    string="Reserve Stock"
+                    help="Pre-book products from stock"
+                    attrs="{'invisible': [('is_stock_reservable', '=', False)]}"
+                />
+            </button>
```
Comment (CasVissers-360ERP) [inline stock_reserve_sale/view/sale.xml:17]: A smartbutton to view reservations would be a nice addition

File: stock_reserve_sale/wizard/sale_stock_reserve.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+        """
+        model = self.env[self.env.context["active_model"]]
+        if model._name == "sale.order":
+            lines = model.browse(self.env.context["active_id"]).order_line
+        else:
+            lines = model.browse(self.env.context["active_ids"])
+
+        try:
+            owners = {line.stock_owner_id for line in lines}
```
Comment (CasVissers-360ERP) [inline stock_reserve_sale/wizard/sale_stock_reserve.py:32]: Fetching owner from order is not working

**Review by CasVissers-360ERP — CHANGES_REQUESTED** [review]
Initial functional review

File: stock_reserve_sale/model/sale.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        return False
+
+    def _get_procure_method(self):
+        """Get procure_method depending on product routes"""
+        rule = self._get_line_rule()
+        if rule:
+            return rule.procure_method
+        return False
+
+    @api.depends("state", "product_id.route_ids", "product_id.type")
```
Comment (CasVissers-360ERP) [inline stock_reserve_sale/model/sale.py:155]: type = detailed_type

File: stock_reserve_sale/model/sale.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    reservation_ids = fields.One2many(
+        "stock.reservation", "sale_line_id", string="Stock Reservation", copy=False
+    )
+    is_stock_reservable = fields.Boolean(
+        compute="_compute_is_stock_reservable", readonly=True, string="Can be reserved"
+    )
+    is_readonly = fields.Boolean(compute="_compute_is_readonly", store=False)
+
+    def release_stock_reservation(self):
+        reserv_ids = [reserv.id for line in self for reserv in line.reservation_ids]
```
Comment (CasVissers-360ERP) [inline stock_reserve_sale/model/sale.py:187]: This can be a lot shorter?

File: stock_reserve_sale/view/sale.xml [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                />
+            </xpath>
+            <field name="user_id" position="before">
+                <field name="has_stock_reservation" invisible="1" />
+            </field>
+            <field name="order_line" position="after">
+                <div class="text-right">
+                    <button
+                        name="release_all_stock_reservation"
+                        string="Cancell all"
```
Comment (CasVissers-360ERP) [inline stock_reserve_sale/view/sale.xml:69]: Cancel all

**Comment by CasVissers-360ERP** [comment]
@flachica The code quality is low, I am going to pause my review.

**Comment by rvalyi** [comment]
Hello, you might consider rebuilding the migration from this v14 migration PR as it bring more bugfix that is not present here https://github.com/OCA/stock-logistics-warehouse/pull/1696

**Pull request closed without merging**